src/sentry/static/sentry/app/bootstrap.tsx

```diff
@@ -70,10 +73,60 @@ const config = ConfigStore.getConfig();
 
 const tracesSampleRate = config ? config.apmSampling : 0;
 
+const appRoutes = Router.createRoutes(routes());
+const createLocation = createMemoryHistory().createLocation;
+
 Sentry.init({
   ...window.__SENTRY__OPTIONS,
   integrations: getSentryIntegrations(),
   tracesSampleRate,
+  async beforeSend(event) {
+    if (event.type === 'transaction') {
+      // for JavaScript transactions, set the transaction name based on the current window.location
+      // object against the application react-router routes
+
+      let prevTransactionName = event.transaction;
+
+      if (typeof prevTransactionName === 'string') {
+        if (prevTransactionName.startsWith('/')) {
+          return event;
+        }
+      } else {
+        prevTransactionName = window.location.pathname;
+      }
+
+      const transactionName: string | undefined = await new Promise(function(resolve) {
+        Router.match(
+          {
+            routes: appRoutes,
+            location: createLocation(prevTransactionName),
+          },
+          (error, _redirectLocation, renderProps) => {
+            if (error) {
```

Comment:
Do we know the cases when this might fail?
Because in those cases we would still get the full url.
I just wanted to highlight this.

Reply:
Unfortunately we may not know. 

We need to at least handle this otherwise the transaction name will be overridden by a worst choice of a transaction name; in this case, a blank string.